src/Native/System.Net.Security.Native/pal_gssapi.cpp

```diff
@@ -0,0 +1,265 @@
+                                                GSS_C_NO_CHANNEL_BINDINGS,
+                                                &inputToken,
+                                                nullptr,
+                                                *outBufferHandle,
+                                                retFlags,
+                                                nullptr);
+
+    NetSecurityNative_HandleError(majorStatus, outBufferHandle, outTokenLength);
+    return majorStatus;
+}
+
+extern "C" uint32_t NetSecurityNative_ReleaseCred(uint32_t* minorStatus, GssCredId** credHandle)
+{
+    assert(minorStatus != nullptr);
+    assert(credHandle != nullptr);
+    return gss_release_cred(minorStatus, credHandle);
+}
+
+extern "C" uint32_t NetSecurityNative_ReleaseBuffer(uint32_t* minorStatus, GssBuffer* buffer)
+{
+    assert(minorStatus != nullptr);
+    assert(buffer != nullptr);
+    uint32_t status = gss_release_buffer(minorStatus, buffer);
+    delete buffer;
+    return status;
+}
+
+extern "C" void
+NetSecurityNative_CopyBuffer(GssBuffer* bufferHandle, uint8_t* bytes, uint32_t capacity, uint32_t offset)
+{
+    assert(bufferHandle != nullptr);
+    printf("capacity = %d, offset = %d, length = %zu \n", capacity, offset, bufferHandle->length);
+    assert(bufferHandle->length <= (capacity - offset));
+    memcpy(bytes + offset, bufferHandle->value, bufferHandle->length);
+}
+
+extern "C" uint32_t NetSecurityNative_ReleaseName(uint32_t* minorStatus, GssName** inputName)
+{
+    assert(minorStatus != nullptr);
+    assert(inputName != nullptr);
+    return gss_release_name(minorStatus, inputName);
+}
+
+extern "C" uint32_t NetSecurityNative_Wrap(uint32_t* minorStatus,
+                                           GssCtxId* contextHandle,
+                                           int32_t isEncrypt,
+                                           uint8_t* inputBytes,
+                                           int32_t offset,
+                                           int32_t count,
+                                           GssBuffer** outBufferHandle,
+                                           uint32_t* outMsgLength)
+{
+    assert(minorStatus != nullptr);
+    assert(contextHandle != nullptr);
+    assert(isEncrypt == 1 || isEncrypt == 0);
+    int confState;
+    gss_buffer_desc inputMessageBuffer{.length = UnsignedCast(count), .value = inputBytes + offset};
+    *outBufferHandle = new GssBuffer();
+    uint32_t majorStatus = gss_wrap(
+        minorStatus, contextHandle, isEncrypt, GSS_C_QOP_DEFAULT, &inputMessageBuffer, &confState, *outBufferHandle);
+    NetSecurityNative_HandleError(majorStatus, outBufferHandle, outMsgLength);
+    return majorStatus;
+}
+
+extern "C" uint32_t NetSecurityNative_Unwrap(uint32_t* minorStatus,
+                                             GssCtxId* contextHandle,
+                                             uint8_t* inputBytes,
+                                             int32_t offset,
```

Comment:
In addition to asserting that any pointers aren't null, do asserts for bounds checks (that is, validate offset and/or count)